expenses/reports.py

```diff
@@ -149,14 +149,40 @@ def tabulate(self, results: typing.Iterable, engine: Engine) -> SafeString:
             )
         )
 
-    def run(self) -> SafeString:
+    def create_file(self, results: typing.Iterable, engine: Engine) -> SafeString:
+        column_headers: (typing.List[str], typing.List[str]) = self.get_column_header_names(engine)
+        column_header_names, _ = column_headers
+
+        first_row, results = peek(self.preprocess_rows(results))
+
+        if not results:
+            return no_results_to_show()
+
+        if len(column_header_names) != len(first_row):
+            raise ValueError("Results do not match expected column headers")
+        
+
+        return mark_safe(
```

Comment:
This function should return a regular string and create a CSV file using the `csv` library, not try produce it by hand with string templating operations. Things get more complex if a `"`, `\`, or newline appears in fields (and it’s possible with free-form inputs, such as the category names.

Also, note that the CSV file produced contains the currency symbol. This isn’t usual in CSV files, not all software will interpret it correctly. Microsoft Excel could in theory interpret it, but it expects a regular space (U+0020) instead of a non-breaking space (U+00A0) that we use. The rendering pipeline assumes HTML a bit too much; this could be fixed by injecting `format_money` etc. as an argument (and pass a different function depending on output format) + an `is-html` argument to avoid creating links in `VendorStats`.